// synchro-studio/routes/dev.js
exports.dev = function * (synchroStudio, appName, req, res)
{
    var apiProcessor = synchroStudio.getApiProcessor(appName);
    var moduleStore = yield synchroStudio.getAppModuleStoreForPathAwaitable(appName);
    if (!moduleStore)
    {
        res.status(404).send('Module "' + appName + '" not found');
        return;
    }

    var page = null;
    var code = "";
    if (req.query["page"])
    {
        page = req.query["page"];
    }

    if (page)
    {
        code = yield moduleStore.getModuleSourceAwaitable(page);
        if (code && Buffer.isBuffer(code))
        {
            code = code.toString();
        }
    }

    var files = yield moduleStore.listModulesAwaitable();
    var debugPort = apiProcessor.debugPort;
    var modulePathPrefix = synchroStudio.config.get('APP_ROOT_PATH') + "/" + apiProcessor.appContainer + "/";

    var loggedInUser = req.session.user_id;

    synchroStudio.render('dev', {
        modulePathPrefix: modulePathPrefix, 
        code: code, 
        page: page, 
        files: files, // !!! modules
        debugPort: debugPort, 
        loggedInUser: loggedInUser
    }, res);
};
