// synchro-studio/routes/edit.js
/*
 * GET edit page.
 */
var path = require('path');
var logger = require('log4js').getLogger("studio-edit");

// GET /edit
exports.edit = function * (synchroStudio, appName, req, res) // Generator
{
    var apiProcessor = synchroStudio.getApiProcessor(appName);
    var moduleStore = yield synchroStudio.getAppModuleStoreForPathAwaitable(appName);
    if (!moduleStore)
    {
        res.status(404).send('Module "' + appName + '" not found');
        return;
    }

    var page = null;
    var code = "";
    if (req.query["page"])
    {
        page = req.query["page"];
    }

    if (page)
    {
        code = yield moduleStore.getModuleSourceAwaitable(page);
        if (code && Buffer.isBuffer(code))
        {
            code = code.toString();
        }
    }

    var files = yield moduleStore.listModulesAwaitable();
    var debugPort = apiProcessor.debugPort;
    var modulePathPrefix = synchroStudio.config.get('APP_ROOT_PATH') + "/" + apiProcessor.appContainer + "/";

    var loggedInUser = req.session.user_id;

    synchroStudio.render('sandbox', { 
        title: 'Synchro', 
        modulePathPrefix: modulePathPrefix, 
        code: code, 
        page: page, 
        files: files, 
        debugPort: debugPort, 
        loggedInUser: loggedInUser
    }, res);
};

// !!! Create app
//
exports.createApp = function * (synchroStudio, req, res)
{
    var appPath;
    var appDescription;
    var appContainer;

    var apiProcessorManager = synchroStudio.getApiManager()
    var modulesStore = apiProcessorManager.createModuleStore();

    try
    {
        // This will error if the container already exsists...
        //
        yield modulesStore.createAppContainerAwaitable(appContainer);
    }
    catch (err)
    {
        // This isn't a problem per se (multiple app installs can point to the same container)
        //
        console.log("Error:", err.message);
    }

    var appModuleStore = yield modulesStore.getAppModuleStoreAwaitable(appContainer);

    // !!! Create package.json and main.js in module store

    // !!! Add app to config

}

// POST /module (create)
//
exports.createModule = function * (synchroStudio, appName, req, res) // Generator
{
    var apiProcessor = synchroStudio.getApiProcessor(appName);
    var moduleStore = yield synchroStudio.getAppModuleStoreForPathAwaitable(appName);

    logger.info("Create module");
    var result = { };

    // !!! Make sure module does not already exist, and if it does, return appropriate result status/message
    //

    if (req.body["module"] && req.body["source"]) // !!! Is empty string OK?  Seems like we should allow it.
    {
        var moduleName = req.body["module"];
        var source = req.body["source"];

        logger.info("Putting module source");
        var putResult = yield moduleStore.putModuleSourceAwaitable(moduleName, source);

        logger.info("Reloading module");
        var reloadResult = yield apiProcessor.reloadModuleAwaitable(moduleName);
        logger.info("Done reloading module, result:", result);
        if (reloadResult["error"]) 
        {
            result.status = "Fail";
            result.message = "Unable to deploy new module, reason: " + reloadResult["error"];
        }
        else
        {
            result.status = "OK";
            result.message = "Module created and deployed";
        } 
    }
    else
    {
        result.status = "Fail";
        result.message = "Module source not saved";
    }

    res.send(result);
};

// GET /module (read)
//
exports.loadModule = function * (synchroStudio, appName, req, res) // Generator
{
    var moduleStore = yield synchroStudio.getAppModuleStoreForPathAwaitable(appName);
    var moduleName = req.query["module"];

    logger.info("Load module: " + moduleName);
    var result = { };

    if (moduleName)
    {
        result.status = "OK";
        result.message = "Module source found";
        result.source = yield moduleStore.getModuleSourceAwaitable(moduleName);
        if (result.source && Buffer.isBuffer(result.source))
        {
            result.source = result.source.toString();
        }
    }
    else
    {
        result.status = "Fail";
        result.message = "No module specified";
    }
    res.send(result);
};

// POST /module (update)
//
exports.saveModule = function * (synchroStudio, appName, req, res) // Generator
{
    var apiProcessor = synchroStudio.getApiProcessor(appName);
    var moduleStore = yield synchroStudio.getAppModuleStoreForPathAwaitable(appName);

    logger.info("Save module");
    var result = { };

    if (req.body["module"] && req.body["source"]) // !!! Is empty string OK?  Seems like we should allow it.
    {
        var moduleName = req.body["module"];
        var source = req.body["source"];

        logger.info("Putting module source");
        var putResult = yield moduleStore.putModuleSourceAwaitable(moduleName, source);

        logger.info("Reloading module");
        var reloadResult = yield apiProcessor.reloadModuleAwaitable(moduleName);
        logger.info("Done reloading module, result:", result);
        if (reloadResult["error"]) 
        {
            result.status = "Fail";
            result.message = "Unable to deploy module, reason: " + reloadResult["error"];
        }
        else
        {
            result.status = "OK";
            result.message = "Module updated and deployed";
        } 
    }
    else
    {
        result.status = "Fail";
        result.message = "Module source not saved";
    }

    res.send(result);
};

// DELETE /module (delete)
//
exports.deleteModule = function * (synchroStudio, appName, req, res) // Generator
{
    var apiProcessor = synchroStudio.getApiProcessor(appName);
    var moduleStore = yield synchroStudio.getAppModuleStoreForPathAwaitable(appName);

    logger.info("Delete module");
    var result = { };

    if (req.body["module"])
    {
        var moduleName = req.body["module"];

        logger.info("Deleting module:", moduleName);
        yield moduleStore.removeModuleSourceAwaitable(moduleName);
        yield apiProcessor.reloadModuleAwaitable();

        result.status = "OK";
        result.message = "Module deleted";
    }
    else
    {
        result.status = "Fail";
        result.message = "No module specified";
    }

    res.send(result);
};
