// synchro-api/test/testapp/counter.js
// Special case of view initializer with no view template defined in exports.View
//
exports.InitializeView = function(context, session, viewModel, view)
{
    var view =
    {
        title: "Counter Page",
        onBack: "exit",
        elements: 
        [
            { control: "text", value: "Count: {count}", font: 24 },
        ]
    };

    return view;
}

exports.InitializeViewModel = function(context, session)
{
    var viewModel =
    {
        count: 0,
    }
    
    // We only want to do this for tests where we have a real API processor (as opposed to module manager / loader
    // tests which call into InitializeViewModel with mock context and session and don't have an api processor.
    //
    if (context.session && context.request && context.response)
    {
        viewModel.custom = Synchro.getConfig(context, "custom");
    }

    return viewModel;
}

exports.Commands =
{
    test: function(context, session, viewModel)
    {
        viewModel.url = Synchro.getResourceUrl("user.png");
    },
    inc: function(context, session, viewModel)
    {
        viewModel.count += 1;
    },
    reset: function(context, session, viewModel)
    {
        viewModel.count = 0;
    },
}
