// synchro-api/test/module-test.js
require('./test');

var assert = require("assert");
require("./assert-helper");

var path = require('path');
var net = require('net');
var nconf = require('nconf');

var synchroApiModule = require("../index");

var logger = require('log4js').getLogger("module-test");

// This is testing the functionality of the top level module APIs (in index.js of this module).

// Test fixture
exports.createService = function(serviceName, serviceConfiguration)
{
    var service;

    switch (serviceName)
    {
        case "TestService":
        {
        	service = 
        	{
        		name: "TestService",
        		config: serviceConfiguration
        	};
        }
        break;
    }

    return service;
}

describe("Synchro API module", function () 
{
	describe("Service creation", function () 
	{
		it("Should create test service from spec", function() 
		{
			var serviceSpec =
			{
				packageRequirePath: path.resolve(__dirname, 'module-test'),
				serviceName: "TestService",
				serviceConfiguration: { foo: "bar"}
			}

			var service = synchroApiModule.createServiceFromSpec(serviceSpec);
			assert.objectsEqual(service, { name: "TestService", config: { foo: "bar"} });
		});
	});

	describe("Processor Manager", function() 
	{
        var conf = new nconf.Provider();
        conf.defaults({
            "APP_ROOT_PATH": path.resolve(__dirname, '../'),
            "SESSIONSTORE_PACKAGE": path.resolve(__dirname, '../index'),
            "SESSIONSTORE_SERVICE": "MemorySessionStore",
            "SESSIONSTORE": {},
            "MODULESTORE_PACKAGE": path.resolve(__dirname, '../index'),
            "MODULESTORE_SERVICE": "FileModuleStore",
            "MODULESTORE": { directory: __dirname },
            "APP_RESOURCE_PREFIX": "test",
            "APPS":  
            {
                "inproc": 
                {
                    "container": "testapp",
                    "custom": "foo"
                },
                "forked":
                {
                    "container": "testapp",
                    "custom": "bar"
                },
                "inproc2": 
                {
                    "container": "testapp"
                },
                "forked2":
                {
                    "container": "testapp"
                },
            }
        });

		var apiProcessorManager = synchroApiModule.createApiProcessorManager(6969, conf);
		var inprocProcessor;
		var forkedProcessor;

		it("Should create API processor manager", function() 
		{
			assert.notEqual(apiProcessorManager, null);
		});

		it("Should create specified in-proc API processor", function(done)
		{
			apiProcessorManager.createApiProcessorAsync("inproc", false, false, function(err, apiProcessor)
			{
				inprocProcessor = apiProcessor;
				assert.equal(err, null);
				assert.notEqual(inprocProcessor, null);
				assert.equal(inprocProcessor.isForked, false);
				done();
			});
		});

		it("Should create specified forked API processor", function(done) 
		{
			apiProcessorManager.createApiProcessorAsync("forked", true, true, function(err, apiProcessor)
			{
				forkedProcessor = apiProcessor;
				assert.equal(err, null);
				assert.notEqual(forkedProcessor, null);
				assert.equal(forkedProcessor.isForked, true);
				done();
			});
		});

		it("Should fail to create API processor if one already exists at the specified path", function(done) 
		{
			apiProcessorManager.createApiProcessorAsync("inproc", false, false, function(err, apiProcessor)
			{
				assert.notEqual(err, null);
				assert.equal(apiProcessor, null);
				done();
			});
		});

		it("Should find proper API processor based on path", function()
		{
			assert.equal(apiProcessorManager.getApiProcessor("inproc"), inprocProcessor);
			assert.equal(apiProcessorManager.getApiProcessor("forked"), forkedProcessor);
		});

		it("Should fail to find API processor when none coorespond to path", function()
		{
			assert.equal(apiProcessorManager.getApiProcessor("foo"), null);
		});

		it("Should return all API processors", function()
		{
			var apiProcessors = apiProcessorManager.getApiProcessors();
			assert.equal(Object.keys(apiProcessors).length, 2);
			assert.equal(apiProcessors["inproc"], inprocProcessor);
			assert.equal(apiProcessors["forked"], forkedProcessor);
		});

		it("Should return module store for API processor", function()
		{
            var moduleStoreSpec =
            {
                packageRequirePath: conf.get('MODULESTORE_PACKAGE'),
                serviceName: conf.get('MODULESTORE_SERVICE'),
                serviceConfiguration: conf.get('MODULESTORE')
            }

            var inprocModuleStore = apiProcessorManager.getAppModuleStore("inproc2", "testapp", moduleStoreSpec);
            var forkedModuleStore = apiProcessorManager.getAppModuleStore("forked2", "testapp", moduleStoreSpec);
            
            assert.notEqual(inprocModuleStore, null);
            assert.notEqual(forkedModuleStore, null);
			assert.equal(apiProcessorManager.getAppModuleStoreForPath("inproc2"), inprocModuleStore); 
			assert.equal(apiProcessorManager.getAppModuleStoreForPath("forked2"), forkedModuleStore); 
		});

		it("Should get proper app definition from API processor", function(done)
		{
			var request = { headers: [], body: { Mode: "AppDefinition" } };
			var response = 
			{
				send: function(data)
				{
					var expectedAppDefinition = 
					{
						App:
						{
							"name": "synchro-test",
							"version": "0.0.0",
							"description": "Synchro API Test",
							"main": "launch",
							"author": "Bob Dickinson <bob@synchro.io> (http://synchro.io/)",
							"private": true,
							"engines": { "synchro": "*" }						
						}
					};

					assert.objectsEqual(data, expectedAppDefinition);
					done();
				}
			};

			apiProcessorManager.processHttpRequest("inproc", request, response);
		});

		it("Should route HTTP request to in-proc API processor", function(done)
		{
			var request = { headers: [], body: { Mode: "Page", Path: "counter" } };
			var response = 
			{
				send: function(data)
				{
					assert.notEqual(data, null);
					assert.equal(data.View.title, "Counter Page");
					assert.equal(data.ViewModel.count, 0);
					done();
				}
			};

			apiProcessorManager.processHttpRequest("inproc", request, response);
		});
        
		it("Should route HTTP request to forked API processor", function(done)
		{
			var request = { headers: [], body: { Mode: "Page", Path: "counter" } };
			var response = 
			{
				send: function(data)
				{
					assert.notEqual(data, null);
					assert.equal(data.View.title, "Counter Page");
					assert.equal(data.ViewModel.count, 0);
					done();
				}
			};

			apiProcessorManager.processHttpRequest("forked", request, response);
        });

        it("Should get custom app config in in-proc API processor", function (done)
        {
            var request = { headers: [], body: { Mode: "Page", Path: "counter" } };
            var response = 
            {
                send: function (data)
                {
                    assert.notEqual(data, null);
                    assert.equal(data.View.title, "Counter Page");
                    assert.equal(data.ViewModel.custom, "foo");
                    done();
                }
            };
            
            apiProcessorManager.processHttpRequest("inproc", request, response);
        });

        it("Should get custom app config in forked API processor", function (done)
        {
            var request = { headers: [], body: { Mode: "Page", Path: "counter" } };
            var response = 
            {
                send: function (data)
                {
                    assert.notEqual(data, null);
                    assert.equal(data.View.title, "Counter Page");
                    assert.equal(data.ViewModel.custom, "bar");
                    done();
                }
            };
            
            apiProcessorManager.processHttpRequest("forked", request, response);
        });
	});

	describe("Built-in Services", function() 
	{
		it("Should create file module store", function() 
		{
			var service = synchroApiModule.createService("FileModuleStore", { directory: __dirname } );
			var appModuleStore = service.getAppModuleStore("testapp");
			var appDefinition = appModuleStore.getAppDefinition();
			assert.equal(appDefinition.name, "synchro-test");
		});

		it("Should create memory session store", function() 
		{
			var service = synchroApiModule.createService("MemorySessionStore", { } );
			var sessionId = service.createSession();
			assert.notEqual(sessionId, null);
		});

		it("Should create resource resolver", function() 
		{
			var service = synchroApiModule.createService("ResourceResolver", { prefix: "test:" } );
			var resourceUrl = service. getResourceUrl("foo");
			assert.equal(resourceUrl, "test:foo");
		});

		it("Should create file session store");
		it("Should create Redis session store");
	});
});
