// synchro-studio/public/script/sandbox.js
// Edit page javascript
//
var currentModulePathPrefix = null;
var currentModule = null; // Only if loaded via modules (not via debugger)
var currentScriptPath = null;
var editMode = false;

var editor;

// Toastr rules.  http://codeseven.github.io/toastr/
//
toastr.options = 
{
    closeButton: true,
    debug: false,
    positionClass: "toast-bottom-right",
    onclick: null,
    showDuration: 300,
    hideDuration: 1000,
    timeOut: 5000,
    extendedTimeOut: 1000,
    showEasing: "swing",
    hideEasing: "linear",
    showMethod: "fadeIn",
    hideMethod: "fadeOut"
}

// Add put and delete AJAX functions
//
//   http://stackoverflow.com/questions/2153917/how-to-send-a-put-delete-request-in-jquery
//
//   http://stepansuvorov.com/blog/2014/04/jquery-put-and-delete/
//
jQuery.each( [ "put", "delete" ], function( i, method ) {
  jQuery[ method ] = function( url, data, callback, type ) {
    if ( jQuery.isFunction( data ) ) {
      type = type || callback;
      callback = data;
      data = undefined;
    }

    return jQuery.ajax({
      url: url,
      type: method,
      dataType: type,
      data: data,
      success: callback
    });
  };
});

// Here is some a discussion and some code to deal with moving breakpoints on line insert/delete:
//
//     https://groups.google.com/forum/#!msg/ace-discuss/sfGv4tRWZdY/ca1LuolbLnAJ
//     https://github.com/MikeRatcliffe/Acebug/blob/master/chrome/content/ace%2B%2B/startup.js
//

function loadSource(sourceData)
{
    // sourceData.moduleName
    // sourceData.scriptPath
    // sourceData.source
    // sourceData.breakpoints
    // sourceData.executionPointer

    currentModule = sourceData.moduleName;
    currentScriptPath = sourceData.scriptPath;

    editor.session.clearBreakpoints();
    editor.session.setValue(sourceData.source);
    if (sourceData.breakpoints)
    {
        for (var i = 0; i < sourceData.breakpoints.length; i++)
        {
            var breakpoint = sourceData.breakpoints[i];
            editor.session.setBreakpoint(breakpoint.line);
        }
    }
    var executionPointer = sourceData.executionPointer || -1;
    setActiveBreakpoint(executionPointer);
    if (executionPointer >= 0)
    {
        editor.renderer.scrollToLine(executionPointer, true, true);
    }

    $("#module").text("Module: " + currentModule);
    if (currentModule)
    {
        $("button#save").show();
    }
    else
    {
        $("button#save").hide();
    }

    // Highlight the new active module...
    //
    $("div#modules a.active").removeClass("active");
    $("div#modules a[module='" + currentModule + "']").addClass("active");
}

function initEditPage(modulePathPrefix, page)
{
    currentModulePathPrefix = modulePathPrefix;
    currentScriptPath = currentModulePathPrefix + page;
    editMode = true;

    editor = ace.edit("editor");
    editor.setTheme("ace/theme/monokai");
    editor.getSession().setMode("ace/mode/javascript");
    editor.commands.addCommand(
    {
        name: 'saveCommand',
        bindKey: {win: 'Ctrl-S',  mac: 'Command-S'},
        exec: function(editor) 
        {
            doSave();
        },
        readOnly: false // false if this command should not apply in readOnly mode
    });

    editor.on("guttermousedown", function(e)
    { 
        var target = e.domEvent.target; 
        if (target.className.indexOf("ace_gutter-cell") == -1) 
            return; 
        if (!editor.isFocused()) 
            return; 
        if (e.clientX > 25 + target.getBoundingClientRect().left) 
            return; 

        var row = e.getDocumentPosition().row 
        if (e.editor.session.getBreakpoints()[row]) 
        {
            clearBreakpoint(currentScriptPath, row);
        }
        else 
        {
            setBreakpoint(currentScriptPath, row);
        }
        e.stop();
    });

    editor.session.on("changeBreakpoint", function(e)
    { 
        // alert("change breakpoint");
    });
}

var activeBreakpointMarker;
function setActiveBreakpoint(row) // Use -1 to clear
{
    if (activeBreakpointMarker)
    {
        editor.getSession().removeMarker(activeBreakpointMarker);
        activeBreakpointMarker = null;
    }

    if (row >= 0)
    {
        Range = require("ace/range").Range;
        var range = new Range(row, 0, row, 100);
        activeBreakpointMarker = editor.getSession().addMarker(range,"active-breakpoint","line", true);
    }
}

function onNewModule()
{
    bootbox.prompt(
    {
        title: "Enter module name for new module",
        callback: function (moduleName) 
        {
            if (moduleName === null)
            {
                // prompt dialog cancelled
                return;
            }

            console.log("New module:", moduleName);

            // Validate that moduleName is composed of legal filename chars (path sep allowed)
            var match = moduleName.match(/^[a-zA-Z0-9_]+(?:[\/][a-zA-Z0-9_]+)*([.][a-zA-Z0-9_]+)?$/);
            console.log("Match:", match);

            if (!match)
            {
                bootbox.alert("Bad filename");
                return;
            }

            var moduleSource = [ 
                'exports.View =',
                '{',
                '    title: "' + moduleName + '",',
                '    elements:',
                '    [',
                '        { control: "text", value: "Sample text control" }',
                '    ]',
                '}',
                '',
                'exports.InitializeViewModel = function(context, session)',
                '{',
                '    var viewModel =',
                '    {',
                '    }',
                '    return viewModel;',
                '}'
                ].join("\n");

            if (!match[1])
            {
                // No file extension, add .js
                console.log("Adding .js extension");
                moduleName += ".js";
            }

            $.post("module", { module: moduleName, source: moduleSource }, function(result)
            {
                //alert("Result: " + JSON.stringify(result, null, 4));
                if (result.status == "OK")
                {
                    loadModules(function(err)
                    {
                        if (err)
                        {
                            toastr.error("Error loading module list after creating new module, details: " + err);
                        }
                        else
                        {
                            loadSource(
                            {
                                moduleName: moduleName,
                                scriptPath: currentModulePathPrefix + moduleName,
                                source: moduleSource
                            });
                            toastr.success(result.message);
                        }
                    });
                }
                else if (result.status == "Fail")
                {
                    toastr.error(result.message);
                }
            })
            .fail(function(jqXHR, textStatus) 
            {
                alert("Fail:" + JSON.stringify(jqXHR, null, 4));
                alert("Module create failed: " + textStatus);
            });
        }
    });
}

// Reload the list of modules and repopulate the modules list (such as after adding/removing a module)
// 
function loadModules(cb)
{
    $.getJSON(".", function(data)
    {
        // Process JSON response
        console.log("loadModules - got app info:", JSON.stringify(data));

        $('#modules').empty();
        for (var i = 0; i < data.modules.length; i++)
        {
            $('#modules').append(
                '<a href="#"  class="list-group-item" module="' + data.modules[i] + '" onClick="return onLoadModule(\'' + data.modules[i] + '\');">' + data.modules[i] + '</a>'
            );
        }
        cb();
    })
    .fail(function(jqxhr, textStatus, error) 
    {
        cb(error);
    });
}

function onLoadModule(moduleName)
{
    var modulePath = currentModulePathPrefix + moduleName;
    $.getJSON("module", { module: moduleName }, function(data)
    {
        // Process JSON response
        console.log("loadModule " + moduleName + ": " + JSON.stringify(data));
        loadSource(
        {
            moduleName: moduleName,
            scriptPath: modulePath,
            source: data.source
        });
        loadBreakpoints(modulePath);
    })
    .fail(function() 
    {
        alert( "loadModule error" );
    });

    return false; // To prevent default click behavior
}

function onBreakpointSource(sourceData)
{
    if (sourceData.scriptPath.lastIndexOf(currentModulePathPrefix) == 0)
    {
        sourceData.moduleName = sourceData.scriptPath.substring(currentModulePathPrefix.length);
        console.log("Loading breakpoint source from module loader for module: " + sourceData.moduleName);
        $.getJSON("module", { module: sourceData.moduleName }, function(data)
        {
            // Process JSON response
            console.log("loadModule " + sourceData.moduleName + ": " + JSON.stringify(data));
            sourceData.source = data.source;
            loadSource(sourceData);
        })
        .fail(function() 
        {
            alert( "loadModule error" );
        });
    }
    else
    {
        loadSource(sourceData);
    }
}

function doSave() 
{
    if (!currentModule)
    {
        alert("No module loaded");
        return;
    }

    $.put("module", { module: currentModule, source: editor.getValue() }, function(result)
    {
        //alert("Result: " + JSON.stringify(result, null, 4));
        if (result.status == "OK")
        {
            toastr.success(result.message);
        }
        else if (result.status == "Fail")
        {
            toastr.error(result.message);
        }
    })
    .fail(function(jqXHR, textStatus) 
    {
        alert("Fail:" + JSON.stringify(jqXHR, null, 4));
        alert("Modue save failed: " + textStatus);
    });
}
