// synchro-web/client/core/device-metrics.js
var logger = require('log4js').getLogger("device-metrics");

var DeviceMetrics = function()
{
    this._clientName = "Synchro Web"
    this._clientVersion = "1.3.0";
    this._deviceClass = DeviceMetrics.DeviceClass.Tablet;

    this._os = "Web";
    this._osName = "Web";
    this._deviceName = "Web Browser";
    this._userAgent = navigator.userAgent;

    this._naturalOrientation = DeviceMetrics.Orientation.Landscape;

    this._scalingFactor = 1.0;
    this._deviceScalingFactor = 1.0; // Doesn't really apply to web (where everything is sized in logic/display pixels)

    this._widthDeviceUnits = screen.width;
    this._heightDeviceUnits = screen.height;

    // Screen dims are in "display pixels", which are, very generally and approximately, 1/96 of an inch.
    //    See: http://stackoverflow.com/questions/21680629/getting-the-physical-screen-dimensions-dpi-pixel-density-in-chrome-on-androi
    //
    this._widthInches = this._widthDeviceUnits / 96;
    this._heightInches = this._heightDeviceUnits / 96;

    if (screen.width < 480)
    {
        this._scalingFactor = screen.width/480;
    }

    // Default set to Tablet above
    var diagonal = Math.sqrt((this._widthInches * this._widthInches) + (this._heightInches * this._heightInches));
    if (diagonal < 5)
    {
        this._deviceClass = DeviceMetrics.DeviceClass.Phone;
    }
    else if (diagonal < 7)
    {
        this._deviceClass = DeviceMetrics.DeviceClass.Phablet;
    }
    else if (diagonal < 9)
    {
        this._deviceClass = DeviceMetrics.DeviceClass.MiniTablet;
    }
}


DeviceMetrics.DeviceClass =
{
    Phone:      "Phone",      // 3.5"-5" phones
    Phablet:    "Phablet",    // 5"-6" big phones (Nokia 1520)
    MiniTablet: "MiniTablet", // 7"-8" mini tablets (iPad Min, Nexus 7, etc)
    Tablet:     "Tablet"      // 9"+ tablets (iPad, Surface, etc)
}

DeviceMetrics.DeviceType =
{
    Phone:  DeviceMetrics.DeviceClass.Phone,
    Tablet: DeviceMetrics.DeviceClass.Tablet
}

DeviceMetrics.Orientation = 
{
    Portrait:  "Portrait",
    Landscape: "Landscape",
}

// Client details
//
DeviceMetrics.prototype.getClientName = function()
{ 
    return this._clientName; 
}
DeviceMetrics.prototype.getClientVersion = function()
{ 
    return this._clientVersion; 
}

// Device details
//
DeviceMetrics.prototype.getOS = function()
{ 
    return this._os; 
}
DeviceMetrics.prototype.getOSName = function()
{ 
    return this._osName; 
}
DeviceMetrics.prototype.getDeviceName = function()
{ 
    return this._deviceName; 
}
DeviceMetrics.prototype.getUserAgent = function()
{ 
    return this._userAgent; 
}

// Device type
//
DeviceMetrics.prototype.getDeviceClass = function()
{ 
    return this._deviceClass; 
}
DeviceMetrics.prototype.getDeviceType = function()
{
    return ((this._deviceClass == DeviceMetrics.DeviceClass.Phone) || (this._deviceClass == DeviceMetrics.DeviceClass.Phablet)) ? DeviceMetrics.DeviceType.Phone : DeviceMetrics.DeviceType.Tablet;
}

DeviceMetrics.prototype.getNaturalOrientation = function()
{ 
    return this._naturalOrientation;
}

DeviceMetrics.prototype.getCurrentOrientation = function()
{
    return DeviceMetrics.Orientation.Landscape;
}

// Physical dimensions of device
//
DeviceMetrics.prototype.getWidthInches = function()
{ 
    return this._widthInches; 
}
DeviceMetrics.prototype.getHeightInches = function()
{ 
    return this._heightInches; 
}

// Logical dimensions of device
//
// "Device Units" is a general term to describe whatever units are used to position and size objects in the target environment.
// In iOS this unit is the "point" (a term Apple uses, not to be confused with a typographic point).  In Android, this unit is
// actually the physical pixel value.  In WinPhone this is the "view pixels" value (a virtual coordinate space).
//
DeviceMetrics.prototype.getWidthDeviceUnits = function()
{ 
    return this._widthDeviceUnits;
}
DeviceMetrics.prototype.getHeightDeviceUnits = function()
{ 
    return this._heightDeviceUnits; 
}

// Device scaling factor is the ratio of device units to physical pixels.  This can be used to determine an appropriately sized 
// image resource, for example.
//
DeviceMetrics.prototype.getDeviceScalingFactor = function()
{ 
    return this._deviceScalingFactor; 
}

// Dimensions of device
//
DeviceMetrics.prototype.getWidthUnits = function()
{ 
    return this._widthDeviceUnits / this._scalingFactor; 
}
DeviceMetrics.prototype.getHeightUnits = function()
{ 
    return this._heightDeviceUnits / this._scalingFactor;
}

// Dimensions of view
//
// See: http://stackoverflow.com/questions/1248081/get-the-browser-viewport-dimensions-with-javascript
//
DeviceMetrics.prototype.getViewWidthUnits = function()
{
    return Math.max(document.documentElement.clientWidth, window.innerWidth || 0) / this._scalingFactor;
}
DeviceMetrics.prototype.getViewHeightUnits = function()
{ 
    return  Math.max(document.documentElement.clientHeight, window.innerHeight || 0) / this._scalingFactor;
}
DeviceMetrics.prototype.getViewWidthInches = function()
{
    return Math.max(document.documentElement.clientWidth, window.innerWidth || 0) / 96;
}
DeviceMetrics.prototype.getViewHeightInches = function()
{ 
    return  Math.max(document.documentElement.clientHeight, window.innerHeight || 0) / 96;
}


// Scaling factor is the ratio of logic units to device units.
//
DeviceMetrics.prototype.getScalingFactor = function()
{ 
    return this._scalingFactor; 
}

DeviceMetrics.prototype.synchroUnitsToDeviceUnits = function(synchroUnits)
{
    return (synchroUnits * this._scalingFactor) + "px";
}

// Font scaling - convert font points (typographic points) to Synchro units
//
DeviceMetrics.prototype.typographicPointsToFontSize = function(points)
{
    // For the Web client, Syncho units are also points...
    //
    return points + "pt";
}


module.exports = DeviceMetrics;
